### PR TITLE
Cache and display build logs

### LINKED ISSUE #2535: Cache build log for current/previous build
When you go to the build log in the dashboard, it shows you what's happening with the current build:

Screen Shot 2021-11-28 at 10 49 36 AM

That's useful, but not as useful as it could be.  What I'd really like is to see is the entire log for the current build get sent on load, and then have it continue when there is more data.

Ideally we would cache the build log and send that first, then stream the rest of the log in real-time.

If that's not doable, we can also consider switching to Web Sockets.

### PR DESCRIPTION
**Pull request by humphd**
Fixes #2535.

This adds a cache to the build log stream, so that when new clients connect, we start by sending the beginning of the log that they missed (if any).  It also allows us to stream the previous build's log, so we can always display the last build that completed.

The basic idea is that when we create the stream for the build process, we also create a `cache` stream buffer, which we fill with the data from the build output.  Then, whenever we need to send that data back again, we can get its contents and return.  This also allows us to remember the previous build's log and send it to whoever wants it later on.

I have tested nothing in this PR (yolo!) and I'm not sure how to test this without running it on staging.  This is just a draft to get things started.

### CODE REVIEW

File: src/api/status/public/js/build-log/check-for-build.js [hunk 1]
```diff
@@ -1,48 +1,47 @@
...
 
-async function finish() {
-  try {
-    await reader.cancel();
-  } finally {
-    build = null;
-  }
-}
+  // Prefer the current build, but fallback to the previous one
+  const build = status.current || status.previous;
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/check-for-build.js:9]: Nit
```suggestion
  const build = status.current ?? status.previous;
```

**Review by Andrewnt219 — APPROVED** [review]
Can't say it works for sure because I cannot run the auto-deployment server, but the logic looks straightforward to me. As I understand, the server now additionally writes to the `build.cache` while streaming the new data. And on response, it writes the cache first before appending new data.

Some logic changes in the client but it should work just fine. The code refactoring is a nice addition 👍. Maybe we will need a date library down the road or just wait for [Temporal](https://tc39.es/proposal-temporal/docs/) lol.

File: src/api/status/public/js/build-log/check-for-build.js [hunk 2]
```diff
@@ -1,48 +1,47 @@
...
+    terminal.write(value);
+
+    // Recursively invoke processLog until `done === true`
+    return reader.read().then(processLog).catch(finish);
+  };
+
+  terminal.clear();
+
+  // Start processing the log output messages
+  reader.read().then(processLog).catch(finish);
```
Comment (DukeManh) [inline src/api/status/public/js/build-log/check-for-build.js:46]: Would this be a bit cleaner. 
```suggestion js
const processLog = () => {
    reader
      .read()
      .then(({ done, value }) => {
        if (done) {
          return finish();
        }
        // Write the data to the terminal
        terminal.write(value);
        // Recursively invoke processLog until `done === true`
        return processLog();
      })
      .catch(finish);
  };

  // Start processing the log output messages
  processLog();
```

**Review by DukeManh — CHANGES_REQUESTED** [review]
Great work 👍 coding in the dark.

**Review by Andrewnt219 — DISMISSED** [review]

**Review by Andrewnt219 — DISMISSED** [review]

File: src/api/status/public/js/build-log/check-for-build.js [hunk 3]
```diff
@@ -1,48 +1,48 @@
...
+          return finish();
+        }
+        // Write the data to the terminal
+        terminal.write(value);
+        // Recursively invoke processLog until `done === true`
+        return processLog();
+      })
+      .catch(finish);
+  };
+
```
Comment (DukeManh) [inline src/api/status/public/js/build-log/check-for-build.js:45]: I removed the `terminal.clear()` in the suggestion. Sorry.

**Review by tpmai22 — DISMISSED** [review]
I think it's good, I learned a lot.

**Review by Andrewnt219 — APPROVED** [review]
LGTM🤞.

**Review by DukeManh — APPROVED** [review]

**Comment by humphd** [comment]
I'm not sure what else to do on this without actually merging it and testing live.  @manekenpix I assume you would have to manually set this up if I merged?  Is there a time that's good for us to try that, so I don't break staging?

**Comment by manekenpix** [comment]
@humphd I'll set up the autodeployment server later today (at 6ish) and launch telescope using this branch for testing.

**Comment by humphd** [comment]
Fixed a bunch of crashes @manekenpix found on staging, rebased and integrated the work that landed recently on the build log header.  I think this is ready (again).

**Comment by humphd** [comment]
I missed a lock file fix that went in with @dbelokon's video changes.  Rebased and updated.

**Review by manekenpix — APPROVED** [review]

**Review by HyperTHD — APPROVED** [review]

**Pull request merged**